core/src/test/java/org/infinispan/persistence/ParallelIterationTest.java

```diff
@@ -136,39 +143,61 @@ private void runIterationTest(Executor executor, final boolean fetchValues,
       final ConcurrentMap<Integer, Integer> entries = new ConcurrentHashMap<>();
       final ConcurrentMap<Integer, InternalMetadata> metadata = new ConcurrentHashMap<>();
       final AtomicBoolean sameKeyMultipleTimes = new AtomicBoolean();
-      final AtomicInteger processed = new AtomicInteger();
-      final AtomicBoolean brokenBarrier = new AtomicBoolean(false);
 
       assertEquals(loader.size(), 0);
       insertData();
 
-      loader.process(null, (marshalledEntry, taskContext) -> {
-         int key = unwrapKey(marshalledEntry.getKey());
+      Flowable<MarshalledEntry<Object, Object>> flowable = Flowable.fromPublisher(loader.publishEntries(null, fetchValues, fetchMetadata));
+      flowable = flowable.doOnNext(me -> {
+         Integer key = unwrapKey(me.getKey());
          if (fetchValues) {
             // Note: MarshalledEntryImpl.getValue() fails with NPE when it's got null valueBytes,
             // that's why we must not call this when values are not retrieved
-            Integer existing = entries.put(key, unwrapValue(marshalledEntry.getValue()));
+            Integer existing = entries.put(key, unwrapValue(me.getValue()));
             if (existing != null) {
                log.warnf("Already a value present for key %s: %s", key, existing);
                sameKeyMultipleTimes.set(true);
             }
          }
-         if (marshalledEntry.getMetadata() != null) {
-            log.tracef("For key %d found metadata %s", key, marshalledEntry.getMetadata());
-            InternalMetadata prevMetadata = metadata.put(key, marshalledEntry.getMetadata());
+         if (me.getMetadata() != null) {
+            log.tracef("For key %d found metadata %s", key, me.getMetadata());
+            InternalMetadata prevMetadata = metadata.put(key, me.getMetadata());
             if (prevMetadata != null) {
                log.warnf("Already a metadata present for key %s: %s", key, prevMetadata);
                sameKeyMultipleTimes.set(true);
             }
          } else {
             log.tracef("No metadata found for key %d", key);
          }
-         processed.incrementAndGet();
-      }, executor, fetchValues, fetchMetadata);
+      });
+
+      TestSubscriber<MarshalledEntry<Object, Object>> subscriber = TestSubscriber.create(0);
+      flowable.subscribe(subscriber);
+
+      int batchsize = 10;
+      // Request over the max amount so we can ensure the subscriber will get notified of completion
```

Comment:
the comment is a little outdated, but we request all entries except the last batch (although if the # of entries was a multiple of 9 it wouldn't do that correctly).

The reason this is done is because the way this work is the first thread to request the entries could return all the entries (depends on fast/slow it is running) and this allows me to request at least the last batch on the main thread - guaranteeing we have processing of entries on at least 2 threads.